modules/imonomyBidAdapter.js

```diff
@@ -0,0 +1,126 @@
+import * as utils from 'src/utils';
+import { registerBidder } from 'src/adapters/bidderFactory';
+
+const BIDDER_CODE = 'imonomy';
+const ENDPOINT = '//b.imonomy.com/openrtb/hb/00000';
+const USYNCURL = '//b.imonomy.com/UserMatching/b/';
+
+export const spec = {
+  code: BIDDER_CODE,
+
+  /**
+  * Determines whether or not the given bid request is valid. Valid bid request must have placementId and hbid
+  *
+  * @param {BidRequest} bid The bid params to validate.
+  * @return boolean True if this is a valid bid, and false otherwise.
+  */
+  isBidRequestValid: bid => {
+    return !!(bid && bid.params && bid.params.placementId && bid.params.hbid);
+  },
+  /**
+  * Make a server request from the list of BidRequests.
+  *
+  * @param {validBidRequests[]} - an array of bids
+  * @return ServerRequest Info describing the request to the server.
+  */
+  buildRequests: validBidRequests => {
+    const tags = validBidRequests.map(bid => {
+      // map each bid id to bid object to retrieve adUnit code in callback
+      let tag = {
+        uuid: bid.bidId,
+        sizes: bid.sizes,
+        trid: bid.transactionId,
+        hbid: bid.params.hbid,
+        placementid: bid.params.placementId
+      };
+
+      // add floor price if specified (not mandatory)
+      if (bid.params.floorPrice) {
+        tag.floorprice = bid.params.floorPrice;
+      }
+
+      return tag;
+    });
+
+    // Imonomy server config
+    const time = new Date().getTime();
+    const kbConf = {
+      ts_as: time,
+      hb_placements: [],
+      hb_placement_bidids: {},
+      hb_floors: {},
+      cb: _generateCb(time),
+      tz: new Date().getTimezoneOffset(),
+    };
+
+    validBidRequests.forEach(bid => {
+      kbConf.hdbdid = kbConf.hdbdid || bid.params.hbid;
```

Comment:
On line 57 `kbConf.hdbdid` will always be undefined at this point, so the `||` will only evaluate to else. Possible change:
`kbConf.hdbdid = bid.params.hbid;`

Same goes for line 58 `kbConf.encode_bid`, the `||` will only evaluate to else since `encode_bid` can only be undefined here. Possible change:
`kbConf.encode_bid = bid.params.encode_bid;`

Reply:
Good day to you!
We made all the changes you requested, except those which related to this one (row 57 & 58).
As you can see, we use cycle here, so kbConf.hdbdid and kbConf.encode_bid will be undefined only on the first iteration.